modules/table/src/integrationTest/java/org/apache/ignite/distributed/ReplicasSafeTimePropagationTest.java

```diff
@@ -0,0 +1,303 @@
+     * <ol>
+     *     <li>Start three nodes and a raft group with three peers.</li>
+     *     <li>Send command with safe time X.</li>
+     *     <li>Stop the leader - the only node that actually do safeTime watermark validation within onBeforeApply.</li>
+     *     <li>Send command with safe time less than X to the new leader and verify that SafeTimeReorderException is thrown.</li>
+     * </ol>
+     */
+    @Test
+    public void testSafeTimeReorderingOnLeaderReElection() throws Exception {
+        // Start three nodes and a raft group with three peers.
+        {
+            cluster = Set.of("node1", "node2", "node3").parallelStream().collect(Collectors.toMap(Function.identity(), PartialNode::new));
+
+            startCluster(cluster);
+        }
+
+        PartialNode someNode = cluster.values().iterator().next();
+
+        RaftGroupService raftClient = someNode.raftClient;
+
+        assertThat(raftClient.refreshLeader(), willCompleteSuccessfully());
+
+        long firstSafeTime = calculateSafeTime(someNode.clock);
+
+        // Send command with safe time X.
+        sendSafeTimeSyncCommand(raftClient, firstSafeTime, false);
+
+        // Stop the leader - the only node that actually do safeTime watermark validation within onBeforeApply.
+        assertNotNull(raftClient.leader());
+
+        PartialNode nodeTopStop = cluster.get(raftClient.leader().consistentId());
+
+        assertNotNull(nodeTopStop);
+
+        nodeTopStop.stop();
+
+        // Select alive raft client
+        Optional<PartialNode> aliveNode = cluster.values().stream().filter(node -> !node.nodeName.equals(nodeTopStop.nodeName)).findFirst();
+
+        assertTrue(aliveNode.isPresent());
+
+        RaftGroupService anotherClient = aliveNode.get().raftClient;
+
+        // Send command with safe time less than previously applied to the new leader and verify that SafeTimeReorderException is thrown.
+        sendSafeTimeSyncCommand(anotherClient, firstSafeTime - 1, true);
+
+        sendSafeTimeSyncCommand(anotherClient, calculateSafeTime(aliveNode.get().clock), false);
+    }
+
+    /**
+     * Test verifies that a leader will reject a command with safeTime less than previously applied within leader restart.
+     * <ol>
+     *     <li>Start two and a raft group with two peer.</li>
+     *     <li>Send command with safe time X.</li>
+     *     <li>Restart the cluster.</li>
+     *     <li>Send command with safe time less than previously applied to the leader before the restart
+     *     and verify that SafeTimeReorderException is thrown.</li>
+     * </ol>
+     */
+    @Test
+    public void testSafeTimeReorderingOnLeaderRestart() throws Exception {
+        // Start two node and a raft group with two peer.
+        {
+            cluster = Set.of("node1", "node2").parallelStream().collect(Collectors.toMap(Function.identity(), PartialNode::new));
```

Comment:
I don't think that invoking a constructor required different thread, what was your motivation?

Reply:
Long story, at the very beginning that was create + start. Fixed.